x-pack/plugins/ml/public/application/services/new_job_capabilities/new_job_capabilities.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Aggregation, Field, NewJobCaps } from '../../../../common/types/fields';
+import { ES_FIELD_TYPES } from '../../../../../../../src/plugins/data/public';
+
+// create two lists, one removing text fields if there are keyword equivalents and vice versa
+export function processTextAndKeywordFields(fields: Field[]) {
+  const keywordIds = fields.filter((f) => f.type === ES_FIELD_TYPES.KEYWORD).map((f) => f.id);
+  const textIds = fields.filter((f) => f.type === ES_FIELD_TYPES.TEXT).map((f) => f.id);
+
+  const fieldsPreferringKeyword = fields.filter(
+    (f) =>
+      f.type !== ES_FIELD_TYPES.TEXT ||
+      (f.type === ES_FIELD_TYPES.TEXT && keywordIds.includes(`${f.id}.keyword`) === false)
+  );
+
+  const fieldsPreferringText = fields.filter(
+    (f) =>
+      f.type !== ES_FIELD_TYPES.KEYWORD ||
+      (f.type === ES_FIELD_TYPES.KEYWORD &&
+        textIds.includes(f.id.replace(/\.keyword$/, '')) === false)
+  );
+
+  return { fieldsPreferringKeyword, fieldsPreferringText };
+}
+
+export class NewJobCapabilitiesServiceBase {
+  public _fields: Field[];
```

Comment:
these should be `protected`

Reply:
Updated in 4e2f344ac831a47b06266517d1dbcd0dbd0f3c48